File: backend/app/schemas/llm_usage.py
"""
Schemas for LLM usage monitoring API.
"""
from datetime import datetime, date
from typing import Optional, List

from pydantic import BaseModel, ConfigDict


class LLMUsageLogResponse(BaseModel):
    """Single LLM usage log entry."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: Optional[int] = None
    school_id: Optional[int] = None
    student_id: Optional[int] = None
    teacher_id: Optional[int] = None
    feature: str
    provider: str
    model: str
    prompt_tokens: Optional[int] = None
    completion_tokens: Optional[int] = None
    total_tokens: Optional[int] = None
    latency_ms: Optional[int] = None
    success: bool
    error_message: Optional[str] = None
    estimated_cost_usd: Optional[float] = None
    created_at: datetime


class FeatureBreakdown(BaseModel):
    """Token usage breakdown by feature."""
    feature: str
    total_calls: int
    total_tokens: int
    prompt_tokens: int
    completion_tokens: int
    avg_latency_ms: Optional[int] = None
    error_count: int = 0


class ModelBreakdown(BaseModel):
    """Token usage breakdown by model."""
    model: str
    provider: str
    total_calls: int
    total_tokens: int


class LLMUsageSummary(BaseModel):
    """Aggregated LLM usage summary."""
    total_calls: int
    successful_calls: int
    failed_calls: int
    total_tokens: int
    total_prompt_tokens: int
    total_completion_tokens: int
    avg_latency_ms: Optional[int] = None
    by_feature: List[FeatureBreakdown]
    by_model: List[ModelBreakdown]


class LLMUsageDailyStats(BaseModel):
    """Daily LLM usage statistics."""
    date: date
    total_calls: int
    total_tokens: int
    prompt_tokens: int
    completion_tokens: int
    error_count: int = 0


class LLMUsageUserBreakdown(BaseModel):
    """LLM usage breakdown by user (for school admin)."""
    user_id: Optional[int] = None
    student_id: Optional[int] = None
    teacher_id: Optional[int] = None
    total_calls: int
    total_tokens: int


class LLMUsageSchoolBreakdown(BaseModel):
    """LLM usage breakdown by school (for super admin)."""
    school_id: Optional[int] = None
    total_calls: int
    total_tokens: int

File: admin-v2/src/lib/api/llm-usage.ts
import { apiClient } from './client';

export interface FeatureBreakdown {
  feature: string;
  total_calls: number;
  total_tokens: number;
  prompt_tokens: number;
  completion_tokens: number;
  avg_latency_ms: number | null;
  error_count: number;
}

export interface ModelBreakdown {
  model: string;
  provider: string;
  total_calls: number;
  total_tokens: number;
}

export interface LLMUsageSummary {
  total_calls: number;
  successful_calls: number;
  failed_calls: number;
  total_tokens: number;
  total_prompt_tokens: number;
  total_completion_tokens: number;
  avg_latency_ms: number | null;
  by_feature: FeatureBreakdown[];
  by_model: ModelBreakdown[];
}

export interface LLMUsageDailyStats {
  date: string;
  total_calls: number;
  total_tokens: number;
  prompt_tokens: number;
  completion_tokens: number;
  error_count: number;
}

export interface LLMUsageSchoolBreakdown {
  school_id: number | null;
  total_calls: number;
  total_tokens: number;
}

export interface LLMUsageUserBreakdown {
  user_id: number | null;
  student_id: number | null;
  teacher_id: number | null;
  user_name: string | null;
  user_email: string | null;
  user_role: string | null;
  school_id: number | null;
  total_calls: number;
  total_tokens: number;
  prompt_tokens: number;
  completion_tokens: number;
}

interface DateRange {
  date_from?: string;
  date_to?: string;
}

export const llmUsageApi = {
  getStats: async (params?: DateRange): Promise<LLMUsageSummary> => {
    const { data } = await apiClient.get<LLMUsageSummary>(
      '/admin/global/llm-usage/stats',
      { params }
    );
    return data;
  },

  getDaily: async (params?: DateRange): Promise<LLMUsageDailyStats[]> => {
    const { data } = await apiClient.get<LLMUsageDailyStats[]>(
      '/admin/global/llm-usage/daily',
      { params }
    );
    return data;
  },

  getBySchool: async (params?: DateRange): Promise<LLMUsageSchoolBreakdown[]> => {
    const { data } = await apiClient.get<LLMUsageSchoolBreakdown[]>(
      '/admin/global/llm-usage/by-school',
      { params }
    );
    return data;
  },

  getByUser: async (params?: DateRange): Promise<LLMUsageUserBreakdown[]> => {
    const { data } = await apiClient.get<LLMUsageUserBreakdown[]>(
      '/admin/global/llm-usage/by-user',
      { params }
    );
    return data;
  },
};
